internal/blocktx/processor.go

```diff
@@ -964,6 +997,58 @@ func (p *Processor) publishTxsToMetamorph(ctx context.Context, txs []store.Trans
 	}
 }
 
+func (p *Processor) calculateMissingMerklePaths(ctx context.Context, txs []store.TransactionBlock) (updatedTxs []store.TransactionBlock, err error) {
+	ctx, span := tracing.StartTracing(ctx, "calculateMissingMerklePaths", p.tracingEnabled, p.tracingAttributes...)
+	defer func() {
+		tracing.EndTracing(span, err)
+	}()
+
+	// gather all transactions with missing merkle paths for each block in a map
+	// to avoid getting all transaction from the same block multiple times
+	blockTxsMap := make(map[string][]store.TransactionBlock, 0)
+
+	for _, tx := range txs {
+		if tx.MerklePath == "" {
+			blockTxsMap[string(tx.BlockHash)] = append(blockTxsMap[string(tx.BlockHash)], tx)
+		} else {
+			updatedTxs = append(updatedTxs, tx)
+		}
+	}
+
+	for _, txs := range blockTxsMap {
+		blockHash := txs[0].BlockHash
+
+		txHashes, err := p.store.GetBlockTransactionsHashes(ctx, blockHash)
+		if err != nil {
+			return nil, errors.Join(ErrFailedToGetBlockTransactions, fmt.Errorf("block hash %s", getHashStringNoErr(blockHash)), err)
+		}
+
+		merkleTree := bc.BuildMerkleTreeStoreChainHash(txHashes)
+
+		for _, tx := range txs {
+			if tx.MerkleTreeIndex == -1 {
+				p.logger.Warn("missing merkle tree index for transaction", slog.String("hash", getHashStringNoErr(tx.TxHash)))
+				continue
+			}
+
+			bump, err := bc.NewBUMPFromMerkleTreeAndIndex(tx.BlockHeight, merkleTree, uint64(tx.MerkleTreeIndex)) // #nosec G115
+			if err != nil {
+				return nil, errors.Join(ErrFailedToCreateBUMP, fmt.Errorf("block hash %s", getHashStringNoErr(blockHash)), err)
+			}
+
+			bumpHex, err := bump.String()
+			if err != nil {
+				return nil, errors.Join(ErrFailedToCreateBUMP, fmt.Errorf("block hash %s", getHashStringNoErr(blockHash)), err)
+			}
+
+			tx.MerklePath = bumpHex
+			updatedTxs = append(updatedTxs, tx)
+		}
+	}
+
+	return updatedTxs, nil
+}
+
```

Comment:
I think this would be better done if we change this struct to pointer everywhere in the code. Therefore, I suggest a separate task and PR for that change